### DISCUSSION TITLE
How are redundant commas due to auto-insertion are avoided?

### DISCUSSION THREAD

**Discussion by eldesh** [cue-lang/cue#2425] [Q&A] [answered]
The following input (foo.cue) seems to be valid as a CUE.

```
$ cat foo.cue
foo: 5
,bar: "bar"
$ cue export foo.cue
{
    "foo": 5, "bar": "bar
    "bar": "bar"
}
```

According to https://cuelang.org/docs/references/spec/#commas,

Since the final token on a line is 5(number), a comma is automatically inserted.
Therefore, wouldn't foo.cue be treated as equivalent to the following text?

```cue
foo: 5
,,bar: "bar"
```

**Comment by slewiskelly** [comment]
No. If the comma is already present, another would not be inserted.

Have you observed this behavior? The example given output doesn't show this (it does show another oddity, though).

**Reply by eldesh** [reply]
Sorry, fixed typo.

```
$ cat foo.cue
foo: 5
,bar: "bar"
$ cue export foo.cue
{
    "foo": 5,
    "bar": "bar"
}
```

**Reply by eldesh** [reply]
> No. If the comma is already present, another would not be inserted.

Where in the language specification is the basis for saying this?

**Reply by slewiskelly** [reply, marked as answer]
> Where in the language specification is the basis for saying this?

It's not stated explicitly, but it would make sense that this is the case. Nevertheless, the language spec says that:

> **may** omit most of these commas using the following rules...

and

> To reflect idiomatic use, examples in this document elide commas using these rules.

Therefore the parser is to handle both the inclusion and ommission of commas.

Furthermore, CUE is a superset of JSON; which requires commas at the end of lines, but, annoyingly, not the last element. CUE relaxes both of these rules. [^1]

[^1]: https://cuelang.org/docs/tutorials/tour/intro/json/

**Reply by eldesh** [reply]
>> It's not stated explicitly, 
>and
>> may omit most of these commas using the following rules...

So the "following rules" are not sufficient to define the rules for automatic comma insertion.
That is what I wanted to confirm. Thank you.
